File: oop/Python/FNN/FNN.py

from numpy import exp, array, random, dot, sum, transpose

# activation function
def af(x):
    return 1 / (1 + exp(-x))  # Sigmoid

# slope of activation function
def saf(x):
    return x * (1 - x) # drivative of Sigmoid

# Forward Neraul Network Class
# iN:  Input Nodes
# hN:  Hidden Nodes
# oN:  Output Nodes
# hL:  hidden Layers
# lR:  learning Rate
# ws:  weights
# bs:  biases

class FNN:
    def __init__(self, iN, hN, oN, hL=4, lR=0.1):
        self.iN = iN
        self.hN = hN
        self.oN = oN
        self.hL = hL
        self.lR = lR
        
        # Initialize weights and biases
        self.ws = []
        self.bs = []
        
        # Input to first hidden layer
        self.ws.append(random.uniform(size=(self.iN, self.hN)))
        self.bs.append(random.uniform(size=(1, self.hN)))
        
        # Hidden layers
        for _ in range(hL - 1):
            self.ws.append(random.uniform(size=(self.hN, self.hN)))
            self.bs.append(random.uniform(size=(1, self.hN)))
        
        # Last hidden layer to output layer
        self.ws.append(random.uniform(size=(self.hN, self.oN)))
        self.bs.append(random.uniform(size=(1, self.oN)))

    def feedforward(self, inputs):
        self.activations = [inputs]
        
        # Calculate activations for each layer
        for i in range(self.hL + 1):
            net_input = dot(self.activations[-1], self.ws[i]) + self.bs[i]
            activation = af(net_input)
            self.activations.append(activation)
        
        return self.activations[-1]

    def backpropagate(self, inputs, expected_output):
        output_error = expected_output - self.activations[-1]
        deltas = [output_error * saf(self.activations[-1])]
        
        # Calculate error deltas for each layer
        for i in reversed(range(self.hL)):
            error = deltas[-1].dot(self.ws[i + 1].T)
            delta = error * saf(self.activations[i + 1])
            deltas.append(delta)
        
        deltas.reverse()
        
        # Update weights and biases
        for i in range(self.hL + 1):
            self.ws[i] += self.activations[i].T.dot(deltas[i]) * self.lR
            self.bs[i] += sum(deltas[i], axis=0, keepdims=True) * self.lR

    def train(self, inputs, expected_output, iterations):
        for _ in range(iterations):
            self.feedforward(inputs)
            self.backpropagate(inputs, expected_output)

    def predict(self, inputs):
        return self.feedforward(inputs)

# Example usage with XOR problem
inputs = array([[0, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 1], [1, 0, 0], [1, 0, 1], [1, 1, 0], [1, 1, 1]])
expected_output = array([[0], [1], [1], [0], [1], [0], [0], [1]])

# Create a neural network with 3 input nodes, 4 hidden layers with 4 nodes each, and 1 output node
fnn = FNN(iN=3, hN=4, oN=1)

# Train the neural network
fnn.train(inputs, expected_output, iterations=10000)

# Test the neural network
output = fnn.predict(inputs)
print("Predicted Output:")
print(output)

File: oop/C++/FNN/FNN.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>
#include <ctime>

using namespace std;

// activation function
double af(double x) {
    return 1.0 / (1.0 + exp(-x)); // Sigmoid
}

// slope of activation function
double saf(double x) {
    return x * (1.0 - x);  // drivative of sigmoid
}

// iN:  input Nodes
// hN:  hidden Nodes
// oN:  output Nodes
// iL:  input Layer
// hL:  hidden Layer
// oL:  output Layer
// wIH: weights Input Hidden
// wHO: weights Hidden Output
// w:   weight
// lR:  learning Rate
// oE:  output Errors
// hE:  hidden Errors
class FNN {
private:
    int iN, hN, oN;
    vector<double> iL, hL, oL;
    vector<vector<double>> wIH, wHO;
    double lR;

public:
    FNN(int iN, int hN, int oN, double lR = 0.1)
        : iN(iN), hN(hN), oN(oN), lR(lR) {
        iL.resize(iN);
        hL.resize(hN);
        oL.resize(oN);

        // Initialize weights with random values
        wIH.resize(iN, vector<double>(hN));
        wHO.resize(hN, vector<double>(oN));

        srand(time(0));
        for (auto& row : wIH)
            for (auto& w : row)
                w = ((double)rand() / RAND_MAX) * 2 - 1; // Random values between -1 and 1

        for (auto& row : wHO)
            for (auto& w : row)
                w = ((double)rand() / RAND_MAX) * 2 - 1; // Random values between -1 and 1
    }

    void feedforward(const vector<double>& inputs) {
        iL = inputs;

        for (int i = 0; i < hN; ++i) {
            hL[i] = 0.0;
            for (int j = 0; j < iN; ++j) {
                hL[i] += iL[j] * wIH[j][i];
            }
            hL[i] = af(hL[i]);
        }

        for (int i = 0; i < oN; ++i) {
            oL[i] = 0.0;
            for (int j = 0; j < hN; ++j) {
                oL[i] += hL[j] * wHO[j][i];
            }
            oL[i] = af(oL[i]);
        }
    }

    void backpropagate(const vector<double>& targets) {
        vector<double> oE(oN);
        for (int i = 0; i < oN; ++i)
            oE[i] = targets[i] - oL[i];

        vector<double> hE(hN);
        for (int i = 0; i < hN; ++i) {
            hE[i] = 0.0;
            for (int j = 0; j < oN; ++j)
                hE[i] += oE[j] * wHO[i][j];
            hE[i] *= saf(hL[i]);
        }

        for (int i = 0; i < hN; ++i)
            for (int j = 0; j < oN; ++j)
                wHO[i][j] += lR * oE[j] * hL[i];

        for (int i = 0; i < iN; ++i)
            for (int j = 0; j < hN; ++j)
                wIH[i][j] += lR * hE[j] * iL[i];
    }

    void train(const vector<double>& inputs, const vector<double>& targets, int epochs) {
        for (int epoch = 0; epoch < epochs; ++epoch) {
            feedforward(inputs);
            backpropagate(targets);
        }
    }

    vector<double> predict(const vector<double>& inputs) {
        feedforward(inputs);
        return oL;
    }
};

int main() {
    FNN fnn(3, 4, 1); // 3 input nodes, 4 hidden nodes, 1 output node

    // Training data for a simple problem (e.g., XOR with an additional input)
    vector<vector<double>> trainingInputs = { {0, 0, 0}, {0, 0, 1}, {0, 1, 0}, {0, 1, 1}, {1, 0, 0}, {1, 0, 1}, {1, 1, 0}, {1, 1, 1} };
    vector<vector<double>> trainingOutputs = { {0}, {1}, {1}, {0}, {1}, {0}, {0}, {1} };

    // Train the neural network
    for (int i = 0; i < trainingInputs.size(); ++i)
        fnn.train(trainingInputs[i], trainingOutputs[i], 10000);

    // Test the neural network
    for (int i = 0; i < trainingInputs.size(); ++i) {
        vector<double> output = fnn.predict(trainingInputs[i]);
        cout << trainingInputs[i][0] << " " << trainingInputs[i][1] << " " << trainingInputs[i][2] << " => " << output[0] << endl;
    }

    return 0;
}

